src/Microsoft.Identity.Client/Instance/Discovery/KnownMetadataProvider.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.Instance.Discovery
+{
+    internal class KnownMetadataProvider : IKnownMetadataProvider
+    {
+        // No need to use a ConcurrentDictionary, because the normal Dictionary is thread safe for read operations
+        private static readonly IDictionary<string, InstanceDiscoveryMetadataEntry> s_knownEntries =
+            new Dictionary<string, InstanceDiscoveryMetadataEntry>();
+
+        private static readonly ISet<string> s_knownEnvironments = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+        static KnownMetadataProvider()
+        {
+            InstanceDiscoveryMetadataEntry publicCloudEntry = new InstanceDiscoveryMetadataEntry()
+            {
+                Aliases = new[] { "login.microsoftonline.com", "login.windows.net", "login.microsoft.com", "sts.windows.net" },
+                PreferredNetwork = "login.microsoftonline.com",
+                PreferredCache = "sts.windows.net"
+            };
+
+            InstanceDiscoveryMetadataEntry cnCloudEntry = new InstanceDiscoveryMetadataEntry()
+            {
+                Aliases = new[] { "login.partner.microsoftonline.cn", "login.chinacloudapi.cn" },
+                PreferredNetwork = "login.partner.microsoftonline.cn",
+                PreferredCache = "login.partner.microsoftonline.cn"
+            };
+
+            InstanceDiscoveryMetadataEntry deCloudEntry = new InstanceDiscoveryMetadataEntry()
+            {
+                Aliases = new[] { "login.microsoftonline.de" },
+                PreferredNetwork = "login.microsoftonline.de",
+                PreferredCache = "login.microsoftonline.de"
+            };
+
+            InstanceDiscoveryMetadataEntry usGovCloudEntry = new InstanceDiscoveryMetadataEntry()
+            {
+                Aliases = new[] { "login.microsoftonline.us", "login.usgovcloudapi.net" },
+                PreferredNetwork = "login.microsoftonline.us",
+                PreferredCache = "login.microsoftonline.us"
+            };
+
+            InstanceDiscoveryMetadataEntry usCloudEntry = new InstanceDiscoveryMetadataEntry()
+            {
+                Aliases = new[] { "login-us.microsoftonline.com" },
+                PreferredNetwork = "login-us.microsoftonline.com",
+                PreferredCache = "login-us.microsoftonline.com"
+            };
+
+            AddToKnownCache(publicCloudEntry);
+            AddToKnownCache(cnCloudEntry);
+            AddToKnownCache(deCloudEntry);
+            AddToKnownCache(usGovCloudEntry);
+            AddToKnownCache(usCloudEntry);
+        }
+
+        private static void AddToKnownCache(InstanceDiscoveryMetadataEntry entry)
```

Comment:
so that this is _never_ called anywhere else in this class (thus breaking the concurrency of the dictionary) can you make this a local method within the static constructor instead of a separately scoped private?  That will truly show the intent that writing is ONLY for the static constructor.

Reply:
Hah, I did not know about this C# feature. Cool.

Reply:
nested/local methods are awesome